File: algorithms/sortings/bubbleSort.js
// Geeksforgeeks problem - https://practice.geeksforgeeks.org/problems/punish-the-students5726/1
function shouldPunish(nums, avg) {
  // swapCount to count how many swaps are required
  let swapCount = 0;
  let status = true;
  for (let i = 0; i < nums.length - 1; i++) {
    for (let j = 1; j < nums.length; j++) {
      if (nums[j] < nums[j-1]) {
        swapCount += 1;
        let temp = nums[j];
        nums[j] = nums[j-1];
        nums[j-1] = temp;
      }
    }
  }
  // Deciding if deducting the marks causes
  // the avg marks to go below avg given
  let marksAvg = nums.reduce((acc, curr) => {
    return acc + curr;
  }, 0)
  if (((marksAvg - swapCount) / nums.length) < avg) {
    status = false;
  }
  return status;
}

let array = [50, 67, 89, 79, 58]
console.log("Unsorted:", JSON.stringify(array));
console.log("Should Punish?", shouldPunish(array, 68));
console.log("Sorted:", JSON.stringify(array));

File: algorithms/sortings/bubble_sort.py
# [10, 5, 2, 12, 8, 9]
# [5, 10, 2, 12, 8, 9]
def bubble_sort(arr):
  for i in range(0, len(arr) - 1):
    swapped = False
    for j in range(0, len(arr) - i - 1):
      if arr[j] > arr[j + 1]:
        arr[j], arr[j + 1] = arr[j + 1], arr[j]
        swapped = True
    if not swapped:
      break
  
# unsorted array
arr = [10, 5, 2, 12, 8, 9]
bubble_sort(arr)
# sorted array
print(arr)